account_invoice_margin/models/account_invoice.py

```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+# © 2017 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from openerp import api, fields, models
+import openerp.addons.decimal_precision as dp
+
+
+class AccountInvoiceLine(models.Model):
+    _inherit = 'account.invoice.line'
+
+    margin = fields.Float(
+        compute='_compute_margin',
+        digits=dp.get_precision('Product Price'),
+        store=True,
+        string='Margin',
+    )
+    margin_percent = fields.Float(
+        string='% Margin',
+        compute='_compute_margin',
+        store=True,
+        readonly=True,
+    )
+    purchase_price = fields.Float(
+        digits=dp.get_precision('Product Price'),
+        string='Cost',
+    )
+
+    @api.multi
+    @api.depends('purchase_price', 'price_subtotal')
+    def _compute_margin(self):
+        for line in self.filtered(lambda x: x.invoice_id.type[:2] != 'in'):
+            tmp_margin = line.price_subtotal - (
+                line.purchase_price * line.quantity)
+            if line.invoice_id.type == 'out_refund':
+                tmp_margin *= -1
+            line.update({
+                'margin': tmp_margin,
+                'margin_percent': (tmp_margin / line.price_subtotal * 100.0 if
+                                   line.price_subtotal else 0.0),
+            })
+
+    @api.onchange('product_id', 'uom_id')
+    def _onchange_product_id_margin(self):
+        if self.invoice_id.type in ['out_invoice', 'out_refund']:
+            purchase_price = self.product_id.standard_price
```

Comment:
I don't think it's necessary : that is the same behaviour as in the Odoo core. (sale_margin module)